Patterns in let statements and function parameters (#1214)
powdr-labs/powdr: book/src/pil/patterns.md

are themselves patterns.

Some examples:

```rust
// This pattern destructures the first function parameter.
let f: (int, int), int -> int = |(a, b), c| (a + c, b);
// Matches a tuple, ignores the second component.
let (x, _) = f((6, 7), 3);
// The match statement typically usse patterns to check for certain values
// but it can also destructure and create new local variables valid inside
// the match arm.
let t = match (x, f((1, x), 2)) {
    (0, _) => 0,
    (1, _) => 7,
    (_, y) => y,
    _ => 9
};
let head: int[] -> int = |x| match x {
    // Matches the first element of a non-empty array and binds it to a local variable.
    [a, ..] => a,
    [] => std::check::panic("Called 'head' on empty array."),
};
```

> Note that PIL does not check that patterns in a match expression are exhaustive.

## (Ir-)refutability

A pattern is refutable if there is a value of the correct type that the pattern does not match.
An example is the pattern `7` since it does not match all integers, or the patten `[x, ..]`, because it
does not match the empty array.

Refutable patterns are fine in match arms, because if the pattern does not match, the evaluator will

> just continue trying the next match arm, but they are disallowed in let statements an ind function

Reviewer: ```suggestion
just continue trying the next match arm, but they are disallowed in let statements and in function
```
